test_bandwidth.py

```diff
@@ -0,0 +1,204 @@
+import os
+import sys
+import time
+import argparse
+from abc import abstractmethod
+
+import torch
+import torch.distributed as dist
+from typing import List
+from prettytable import PrettyTable
+os.environ['CUDA_VISIBLE_DEVICES'] = '1,2,3,4'
+    
+def setup(rank, world_size):
+    os.environ["MASTER_ADDR"] = os.getenv("MASTER_ADDR", "localhost")
+    os.environ["MASTER_PORT"] = os.getenv("MASTER_PORT", "12355")
+    os.environ["RANK"] = os.getenv("RANK", "0")
+    os.environ["WORLD_SIZE"] = os.getenv("WORLD_SIZE", "1")
+    dist.init_process_group(world_size=world_size, rank=rank,
+                            init_method="env://", backend="nccl")
+    torch.cuda.set_device(rank)
+    device_id = torch.cuda.current_device()
+    print("Process running on GPU:", device_id)
+
+"""    
+    BandWidth Computation: https://github.com/NVIDIA/nccl-tests/blob/master/doc/PERFORMANCE.md
+
+    B: Bus bandWidth of the device
+    S: Elements transformed when execute one collective op
+    t: Time cost when execute one collective op 
+    n: Ranks number
+
+    AllReduce:
+        B = S/t * (2*(n-1)/n) = algbw * (2*(n-1)/n)
+    AllGather:
+        B = S/t * (n-1)/n = algbw * (n-1)/n
+    Broadcast:
+        B = S/t = algbw 
+    AllToAll:
+        B = S/t * (n-1)/n = algbw * (n-1)/n
+"""
+
+class CollectiveOps:
+    @abstractmethod
+    def __init__(self, world_size: int):
+        self.world_size = world_size
+
+    @abstractmethod
+    def __call__(self, tensor: torch.Tensor) -> None:
+        pass
+
+class AllReduce(CollectiveOps):
+    def __call__(self, tensor: torch.Tensor) -> None:
+        tensor = tensor.contiguous()
+        dist.all_reduce(tensor)
+    
+    def ranks_factor(self):
+        return 2 * (self.world_size - 1) / self.world_size
```

Comment:
if every subclass of `CollectiveOps` implements `ranks_factor`, shouldn't it be an abstract method in `CollectiveOps`? Meanwhile, what does `ranks_factor` mean?

Reply:
Yes, sure. I forgot it. 
In the above comments regarding bandwidth computation, 'n' represents the number of ranks, and all factors are totally about 'n'."